Add YAML viewing functionality for resources.

samzong opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #5986

<img width="3696" height="1978" alt="image" src="https://github.com/user-attachments/assets/557cab38-96b7-40be-bf34-c5f0880b93e9" />


#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
KueueViz: Add View YAML
```

mbobrovskyi commented on cmd/kueueviz/backend/README.md:
```diff
@@ -63,3 +63,9 @@ KUEUE_VIZ_PORT=8181 ./kueue_ws_app
+
+| Endpoint                                          | Description                             |
+| ------------------------------------------------- | --------------------------------------- |
+| `/api/yaml/{resourceType}/{name}?namespace={ns}` | Returns YAML content for a specific resource |
```
Maybe like this?

```suggestion
| `/api/{resource_type}/{namespace}/{name}?output={output_type}` | Returns content for a specific resource |
```

WDYT?

mimowo replied:
+1, especially as we may introduce similar API for showing `describe` (similar to k9s), or `json` outputs.

samzong replied:
This is a good idea, I will optimize it later.

samzong replied:
Explain why REST.

Initially, my design was based on websocket. But, I considered that viewing YAML is more of a one-time fetch action, and the YAML content might be large. Maybe REST is more appropriate to handle this scenario.

as @mimowo say, I'm indeed considering adding more capabilities like View JSON and etc in the future. (By the way, k9s is great.)

And I want to ask how to consider adding create/update capabilities in the future.

mbobrovskyi commented on cmd/kueueviz/frontend/src/ViewYamlButton.jsx:
Could you please add e2e tests to confirm that it works as expected?

mbobrovskyi commented on cmd/kueueviz/backend/main.go:
```diff
@@ -20,36 +20,120 @@ import (
-			log.Fatalf("Error starting pprof server: %v", err)
-		}
-	}()
+	if gin.Mode() != gin.ReleaseMode {
+		go func() {
+			log.Println("Starting pprof server on localhost:6060")
+			if err := http.ListenAndServe("localhost:6060", nil); err != nil {
+				log.Printf("Error starting pprof server: %v", err)
+			}
+		}()
+	}
```
Is it related changes? Can we move it to separate PR?

mbobrovskyi commented on cmd/kueueviz/backend/main.go:
```diff
@@ -20,36 +20,120 @@ import (
-			log.Fatalf("Error starting pprof server: %v", err)
-		}
-	}()
+	if gin.Mode() != gin.ReleaseMode {
```
I’m wondering if we need to add GIN_MODE variable to the manifests.

samzong replied:
👌

mbobrovskyi replied:
When building the YAML manifests for the release, I think we need to add this variable – perhaps [here](https://github.com/kubernetes-sigs/kueue/blob/50e4ba6a497298b1a5523f62ebbb9d5da85422fb/config/kueueviz/kustomization.yaml). But this is not a related change, so we can move it to a follow-up PR.

mbobrovskyi commented on cmd/kueueviz/frontend/src/ViewYamlButton.jsx:
```diff
@@ -0,0 +1,56 @@
+  );
+};
+
+export default ViewYamlButton;
```
```suggestion
export default ViewYamlButton;

```
nit

mbobrovskyi commented on cmd/kueueviz/backend/handlers/api_handlers.go:
```diff
@@ -0,0 +1,88 @@
+		var resource any
+		var fetchErr error
+
+		if namespace != "" {
+			// Namespaced resource
+			resource, fetchErr = dynamicClient.Resource(gvr).Namespace(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+		} else {
+			// Cluster-scoped resource
+			resource, fetchErr = dynamicClient.Resource(gvr).Get(context.TODO(), name, metav1.GetOptions{})
+		}
```
```suggestion
		resource, fetchErr = dynamicClient.Resource(gvr).Namespace(namespace).Get(context.TODO(), name, metav1.GetOptions{})
```

samzong commented on cmd/kueueviz/frontend/src/YamlViewer.jsx:
```diff
@@ -0,0 +1,190 @@
+    setYamlContent('');
+
+    try {
+      const backendUrl = env.REACT_APP_WEBSOCKET_URL || env.VITE_WEBSOCKET_URL;
+      if (!backendUrl) {
+        throw new Error("Backend URL is not configured. Please set REACT_APP_WEBSOCKET_URL or VITE_WEBSOCKET_URL.");
+      }
+
+      const httpUrl = backendUrl.replace('ws://', 'http://').replace('wss://', 'https://');
```
Ok I'll handle it

samzong commented:
@mbobrovskyi @mimowo According your suggestions, I have perfected this PR.

- add `package-lock.json`
- Update the API format and explain why REST was introduced
- add e2e to ViewYamlButton (now, it's passed)
- add `cmd/kueueviz/frontend/src/utils/urlHelper.js`, Extracted public functions and updated websocket and rest
- split `cors.go` and `server.go` from `main.go`, make it to standard go-style
- Optimize `resource, fetchErr :=`

samzong commented on cmd/kueueviz/frontend/src/utils/urlHelper.js:
```diff
@@ -0,0 +1,104 @@
+ * @param {string} wsUrl - The WebSocket URL
+ * @returns {string} The HTTP/HTTPS URL
+ */
+export const convertWebSocketToHttp = (wsUrl) => {
+  return wsUrl.replace("ws://", "http://").replace("wss://", "https://");
+};
```
Good point!

My idea is that the architecture is simpler, because this service port is usually shared, so users only need to configure it once and make a simple conversion. If they configure the same port twice, they will also be troubled.

Of course, separate maintenance is also beneficial. The configuration is more clear. If the subsequent architecture changes, different hosts and ports can be supported.

What's your preference?

samzong replied:
@mbobrovskyi Need u reply 😊

mbobrovskyi replied:
OK. Let's keep it as it is for now.

samzong commented on cmd/kueueviz/frontend/src/utils/urlHelper.js:
```diff
@@ -0,0 +1,104 @@
+ * @throws {Error} If no backend URL is configured
+ */
+export const getBackendWebSocketUrl = () => {
+  const backendUrl = env.REACT_APP_WEBSOCKET_URL || env.VITE_WEBSOCKET_URL;
```
@mbobrovskyi  same as above

mbobrovskyi replied:
OK, let's keep both for now. This is not related to the current PR.

mbobrovskyi commented:
/lgtm
Thank you so much!

samzong commented:
@mbobrovskyi I fixed a `golangci-lint` error, please help review it.

You can see the change at below.

https://github.com/kubernetes-sigs/kueue/compare/142ae0d1281cb3d2764d7e66abb080de0988e3e2..fdfeda581bf480ff08a5c33f0100513fd588685e

mimowo commented:
/approve
